File: src/lib/crawlers/bizinfo.ts
import { extractDate } from '@/lib/utils';
import { BaseCrawler } from './base';
import type { CheerioAPI, AnyNode, Cheerio } from './base';
import type { RawArticle } from './types';

const BASE_URL = 'https://www.bizinfo.go.kr';
const LIST_URL = `${BASE_URL}/web/lay1/bbs/S1T122C128/AS/74/list.do`;

export class BizinfoCrawler extends BaseCrawler {
  name = '기업마당';

  async crawl(): Promise<RawArticle[]> {
    const html = await this.fetch(LIST_URL);
    if (!html) return [];

    const $ = this.parse(html);
    const articles: RawArticle[] = [];

    const rows = this.findRows($);
    for (const row of rows) {
      const article = this.parseRow($, row);
      if (article) articles.push(article);
    }

    console.log(`[${this.name}] ${articles.length}건 수집 완료`);
    return articles;
  }

  private findRows($: CheerioAPI): AnyNode[] {
    const selectors = [
      'table.tbl_type1 tbody tr',
      'table.boardList tbody tr',
      'table.bbs_list tbody tr',
      'div.board_list table tbody tr',
      'div.tbl_wrap table tbody tr',
      'table tbody tr',
    ];
    for (const sel of selectors) {
      const rows = $(sel).toArray();
      if (rows.length > 0) return rows;
    }
    return [];
  }

  private parseRow($: CheerioAPI, row: AnyNode): RawArticle | null {
    const $row = $(row);
    const cols = $row.find('td');
    if (cols.length < 2) return null;

    const link = $row.find('a').first();
    if (!link.length) return null;

    const title = link.text().trim();
    if (!title) return null;

    const url = this.extractUrl(link);
    const date = extractDate($row.text());

    return { title, url, source: this.name, date, deadline: '' };
  }

  private extractUrl(link: Cheerio<AnyNode>): string {
    const href = link.attr('href') ?? '';
    if (href.startsWith('http')) return href;
    if (href.startsWith('/')) return BASE_URL + href;

    const onclick = link.attr('onclick') ?? '';
    if (onclick) {
      const pblancMatch = onclick.match(/PBLN_\w+/);
      if (pblancMatch) {
        return `${BASE_URL}/web/lay1/bbs/S1T122C128/AS/74/view.do?pblancId=${pblancMatch[0]}`;
      }
      const pathMatch = onclick.match(/'(\/[^']+)'/);
      if (pathMatch) {
        return BASE_URL + pathMatch[1];
      }
    }

    return LIST_URL;
  }
}

File: src/crawlers/bizinfo.py
"""기업마당 (bizinfo.go.kr) 지원사업 크롤러."""

from __future__ import annotations

import logging
import re
from datetime import datetime

from bs4 import BeautifulSoup

from .base import Article, BaseCrawler

logger = logging.getLogger(__name__)

BASE_URL = "https://www.bizinfo.go.kr"
LIST_URL = f"{BASE_URL}/web/lay1/bbs/S1T122C128/AS/74/list.do"


class BizinfoCrawler(BaseCrawler):
    """기업마당 지원사업 통합검색 크롤러."""

    name = "기업마당"

    def crawl(self) -> list[Article]:
        resp = self.fetch(LIST_URL)
        if resp is None:
            return []

        resp.encoding = "utf-8"
        soup = BeautifulSoup(resp.text, "html.parser")
        articles: list[Article] = []

        rows = self._find_rows(soup)
        for row in rows:
            article = self._parse_row(row)
            if article:
                articles.append(article)

        logger.info("[%s] %d건 수집 완료", self.name, len(articles))
        return articles

    def _find_rows(self, soup: BeautifulSoup) -> list:
        selectors = [
            "table.tbl_type1 tbody tr",
            "table.boardList tbody tr",
            "table.bbs_list tbody tr",
            "div.board_list table tbody tr",
            "div.tbl_wrap table tbody tr",
            "table tbody tr",
        ]
        for sel in selectors:
            rows = soup.select(sel)
            if rows:
                return rows
        return []

    def _parse_row(self, row) -> Article | None:
        cols = row.select("td")
        if len(cols) < 2:
            return None

        link = row.select_one("a")
        if not link:
            return None

        title = link.get_text(strip=True)
        if not title:
            return None

        url = self._extract_url(link)
        date = self._extract_date(row)

        return Article(title=title, url=url, source=self.name, date=date)

    def _extract_url(self, link) -> str:
        href = link.get("href", "")
        if href and href.startswith("http"):
            return href
        if href and href.startswith("/"):
            return BASE_URL + href

        onclick = link.get("onclick", "")
        if onclick:
            # pblancId 추출
            match = re.search(r"PBLN_\w+", onclick)
            if match:
                return f"{BASE_URL}/web/lay1/bbs/S1T122C128/AS/74/view.do?pblancId={match.group(0)}"
            match = re.search(r"'(/[^']+)'", onclick)
            if match:
                return BASE_URL + match.group(1)

        return LIST_URL

    def _extract_date(self, element) -> str:
        text = element.get_text()
        match = re.search(r"(\d{4})[.\-/](\d{1,2})[.\-/](\d{1,2})", text)
        if match:
            y, m, d = match.groups()
            try:
                dt = datetime(int(y), int(m), int(d))
                return dt.strftime("%Y-%m-%d")
            except ValueError:
                pass
        return datetime.now().strftime("%Y-%m-%d")
